core/blockchain.go

```diff
@@ -1228,6 +1232,29 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, []
 			bc.reportBlock(block, nil, ErrBlacklistedHash)
 			return it.index, events, coalescedLogs, ErrBlacklistedHash
 		}
+		// If the block is known (in the middle of the chain), it's a special case for
+		// Clique blocks where they can share state among each other, so importing an
+		// older block might complete the state of the subsequent one. In this case,
+		// just skip the block (we already validated it once fully (and crashed), since
+		// its header and body was already in the database).
+		if err == ErrKnownBlock {
+			log.Debug("Inserted known block", "number", block.Number(), "hash", block.Hash(),
+				"uncles", len(block.Uncles()), "txs", len(block.Transactions()), "gas", block.GasUsed(),
+				"root", block.Root())
+
+			if err := bc.writeKnownBlock(block); err != nil {
+				return it.index, nil, nil, err
+			}
+			stats.processed++
+
+			// TODO(karalabe): Can we assume canonicalness here? Can we assume no logs?
```

Comment:
The logs are annoying here. Essentially we should fire new logs events if exists. But known block insertion is mainly caused by `Rollback` and `Re-import`. In the `Rollback` function, we didn't fire some events to notify users these logs are removed, so if we fire these logs again, probably users can receive 2 notifications.

Reply:
Perhaps we can judge whether it's a canonical one by total difficulty comparsion.

Reply:
I guess if we're importing a long sidechain (known blocks can only occur with side chains, or lost canons), then we can assume that the TD beat our canon TD, so we can accumulate as canonical.